apollo-router/src/spec/query.rs

```diff
@@ -261,31 +263,54 @@ impl Query {
         vec![]
     }
 
-    pub(crate) fn parse(
-        query: impl Into<String>,
+    fn parse_inner(
+        query: String,
         schema: &Schema,
         configuration: &Configuration,
+        validate: bool,
     ) -> Result<Self, SpecError> {
-        let query = query.into();
         let mut compiler = ApolloCompiler::new()
             .recursion_limit(configuration.preview_operation_limits.parser_max_recursion)
             .token_limit(configuration.preview_operation_limits.parser_max_tokens);
+        compiler.set_type_system_hir(schema.type_system.clone());
         let id = compiler.add_executable(&query, "query");
-        let ast = compiler.db.ast(id);
 
+        let ast = compiler.db.ast(id);
         // Trace log recursion limit data
         let recursion_limit = ast.recursion_limit();
         tracing::trace!(?recursion_limit, "recursion limit data");
 
-        let errors = ast
-            .errors()
-            .map(|err| format!("{err:?}"))
-            .collect::<Vec<_>>();
+        let mut parse_errors = ast.errors().peekable();
+        if parse_errors.peek().is_some() {
+            let text = parse_errors
+                .map(|err| err.to_string())
+                .collect::<Vec<_>>()
+                .join("\n");
+            failfast_debug!("parsing error(s): {}", text);
+            return Err(SpecError::ParsingError(text));
+        }
 
-        if !errors.is_empty() {
-            let errors = errors.join(", ");
-            failfast_debug!("parsing error(s): {}", errors);
-            return Err(SpecError::ParsingError(errors));
+        // Bail out on validation errors, only if the input is expected to be valid
+        if validate && configuration.experimental_graphql_validation != GraphQLValidation::Legacy {
+            let diagnostics = compiler.db.validate_executable(id);
+            let errors = diagnostics
+                .into_iter()
+                .filter(|err| err.data.is_error())
+                .collect::<Vec<_>>();
+
+            for d in &errors {
+                eprintln!("{d}")
```

Comment:
I'm not sure it is a good idea to print directly to stderr. Could you use the same approach as schema validation, where we detect if there's a tty?